usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -560,6 +560,72 @@ static CK_RV check_key_attributes(STDLL_TokData_t * tokdata,
     return rc;
 }
 
+static CK_RV override_key_attributes(STDLL_TokData_t *tokdata,
+                                     CK_KEY_TYPE kt, CK_OBJECT_CLASS kc,
+                                     CK_ATTRIBUTE_PTR attrs,
+                                     CK_ULONG attrs_len)
+{
+    CK_ULONG i;
+    CK_BBOOL cktrue = TRUE;
+    CK_ULONG override_types_public_key[] =
+        { CKA_VERIFY, CKA_ENCRYPT };
+    CK_ULONG *override_types = NULL;
+    CK_BBOOL *override_values[] = { &cktrue, &cktrue };
+    CK_ULONG attr_cnt = 0;
+
+    UNUSED(tokdata);
+
+    switch (kc) {
+    case CKO_PUBLIC_KEY:
+        /*
+         * EP11 does not allow to restrict public RSA/DSA/EC keys with
+         * CKA_VERIFY=FALSE and/or CKA_ENCRYPT=FALSE since it can not
+         * technically enforce the restrictions. Therefore override these
+         * attributes for the EP11 library, but keep the original attribute
+         * values in the object.
+         */
+        if ((kt == CKK_EC) || (kt == CKK_RSA) || (kt == CKK_DSA)) {
+            override_types = &override_types_public_key[0];
+            attr_cnt = sizeof(override_types_public_key) / sizeof(CK_ULONG);
+        }
+        break;
+
+    default:
+        return CKR_OK;
+    }
```

Comment:
i dont get the above switch statement: i mean in case CKO_PUBLIC_KEY, but kt is not rsa, ec, dsa .. in that case loop below does nothing and the case is no different from the default case..

So why not check for all cases when theres nothing to do at the top and return directly ..
```
if (kc != CKO_PUBLIC_KEY || (kt != CKK_ECC && kt != CKK_RSA && kt != CKK_DSA))
    return CKR_OK;
```
.. such that the rest of the code is free of branching.

Reply:
Well, yes, at the moment there is only one switch case, so the whole thing could be done as you suggested. Nevertheless, I would like to keep the switch so that other key classes can be added at a later time easily. 

However, I will change the key type check and return CKR_OK if its not one of the types to handle.